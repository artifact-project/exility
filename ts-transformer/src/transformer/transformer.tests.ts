import * as ts from 'typescript';
import transformer from './transformer';

const compilerOptions = {
	module: ts.ModuleKind.CommonJS,
	target: ts.ScriptTarget.ES5,
};

const printer = ts.createPrinter({
    newLine: ts.NewLineKind.LineFeed,
	removeComments: false,
});

function transform(source: string): string {
	const sourceFile = ts.createSourceFile('source.ts', source, ts.ScriptTarget.Latest, true, ts.ScriptKind.TS);
	const transformationResult = ts.transform(sourceFile, [transformer()], compilerOptions);

	return printer.printFile(transformationResult.transformed[0]);
}


it('string', () => {
	expect(transform(`
		function () {
			class Str extends Block {
				static template = 'h1 | \${attrs.text}!';
			}
		}
	`)).toMatchSnapshot();
});

it('arrow function', () => {
	expect(transform(`
		class ArrowFunction extends Block<null> {
			static template = (attrs) => \`h1 | \${attrs.text}!\`;
		}
	`)).toMatchSnapshot();
});

it('arrow function: if', () => {
	expect(transform(`
		class ArrowIf extends Block<null> {
			static template = (attrs) => \`if (\${attrs.text}) > h1\`;
		}
	`)).toMatchSnapshot();
});

it('arrow function: for (item in items)', () => {
	expect(transform(`
		class ArrowFor extends Block<null> {
			static template = (attrs) => \`for (item in \${attrs.items}) > div | \${item}\`;
		}
	`)).toMatchSnapshot();
});

it('blocks', () => {
	expect(transform(`
		class App extends Block<{username: string}> {
			static blocks = {Button};
			static template = \`
				Button[
					value="Foo"
				]
			\`;
		}
	`)).toMatchSnapshot();
});

it('classNames', () => {
	expect(transform(`
		class App extends Block<{username: string}> {
			static classNames = css;
			static template = \`.foo | bar\`;
		}
	`)).toMatchSnapshot();
});

it('factory', () => {
	expect(transform(`
		function factory<R>(): any {
			class Formify extends Block<R> {
				static template = 'h1 | Wow';
			}
			
			return Formify;
		}
	`)).toMatchSnapshot();
});
