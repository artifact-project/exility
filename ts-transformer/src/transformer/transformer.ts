import * as ts from 'typescript';
import {createCompiler} from '@exility/dom';

function createImport(name, path) {
	return ts.createImportDeclaration(
		undefined,
		undefined,
		ts.createImportClause(ts.createIdentifier(name), undefined),
		ts.createLiteral(path)
	);
}

function addImport(imports: object, path) {
	const name = path.replace(/[^a-z0-9]/g, '_');

	if (!imports.hasOwnProperty(name)) {
		imports[name] = createImport(name, path);
	}

	return imports[name];
}

function generateDepsObject(imports, deps) {
	const props = Object.keys(deps).map(name => {
		const importDecl = addImport(imports, deps[name]);
		const importName = ts.getGeneratedNameForNode(importDecl);

		return ts.createPropertyAssignment(name, ts.createPropertyAccess(importName, 'default'));
	});

	return ts.createObjectLiteral(props);
}


function isBlocks(node) {
	return (
		node.kind === ts.SyntaxKind.PropertyDeclaration &&
		node.name.text === 'blocks' &&
		node.initializer && (node.initializer.kind === ts.SyntaxKind.ObjectLiteralExpression)
	);
}

function isClassNames(node) {
	return (
		node.kind === ts.SyntaxKind.PropertyDeclaration &&
		node.name.text === 'classNames'
	);
}

function hasExtends(node) {
	return node && node.heritageClauses && node.heritageClauses.length;
}

function isStringTemplate({kind}) {
	return (
		kind === ts.SyntaxKind.StringLiteral ||
		kind === ts.SyntaxKind.TemplateExpression ||
		kind === ts.SyntaxKind.NoSubstitutionTemplateLiteral
	);
}

function isArrowTemplate({kind}) {
	return kind === ts.SyntaxKind.ArrowFunction;
}

function isTemplate(node) {
	return (
		node.kind === ts.SyntaxKind.PropertyDeclaration &&
		node.name.text === 'template' &&
		(isStringTemplate(node.initializer) || isArrowTemplate(node.initializer)) &&
		hasExtends(node.parent)
	);
}

function glueTemplateExpression(node: ts.TemplateExpression): string {
	const chunks = [node.head.text];

	node.templateSpans.forEach((node) => {
		const expression = node.expression.getText();
		const code = node.literal.text;

		if (/\b(if\s*\(|in\s*)$/.test(chunks[chunks.length - 1])) {
			chunks.push(`${expression}${code}`);
		} else {
			chunks.push(`\${${expression}}${code}`);
		}
	});

	return chunks.join('');
}

function visitNode(node, imports) {
	if (!isTemplate(node)) {
		return node;
	}

	let templateString = '';

	if (isArrowTemplate(node.initializer)) {
		templateString = glueTemplateExpression(node.initializer.body);
	} else {
		templateString = node.initializer.text;
	}

	const meta = getMeta(node.parent);
	const scope = [
		'attrs',
		'__blocks__',
		'__this__',
		'__slots__',
	];

	meta.cssModule && scope.push('__classNames__');

	const compile = createCompiler({
		...meta,
		scope,
	});
	const templateFactory = compile(templateString);
	const code = templateFactory
					.toString()
					.replace(/__STDDOM_CMP_SET_COMPILER\(.*?\);\n/, '');

	return ts.updateProperty(
		node,
		undefined,
		node.modifiers,
		ts.createIdentifier('prototype.__template__'),
		node.type,
		ts.createCall(
			ts.createParen(ts.createIdentifier(code)),
			undefined,
			[generateDepsObject(imports, {
				'stdlib': '@exility/compile-helpers/src/stdlib/stdlib',
				'stddom': '@exility/dom/src/stddom/stddom',
			})]
		)
	);
}

function visitNodeAndChildren(node, context, imports) {
	// Ходим только по классам, методам, функциям и свойствам
	if (
		node == null || !(
			node.kind === ts.SyntaxKind.SourceFile ||
			node.kind === ts.SyntaxKind.PropertyDeclaration ||
			node.kind === ts.SyntaxKind.ClassDeclaration ||
			node.kind === ts.SyntaxKind.ClassExpression ||
			node.kind === ts.SyntaxKind.ArrowFunction ||
			node.kind === ts.SyntaxKind.FunctionDeclaration ||
			node.kind === ts.SyntaxKind.FunctionExpression
		)
	) {
		return node;
	}

	return ts.visitEachChild(
		visitNode(node, imports),
		(childNode) => visitNodeAndChildren(childNode, context, imports),
		context
	);
}

function getMeta(node) {
	const meta = {blocks: [], cssModule: false};

	function visitNode(node) {
		if (isBlocks(node)) {
			meta.blocks = node.initializer.properties.map(node => node.name.text);
		} else if (isClassNames(node)) {
			meta.cssModule = true;
		}
	}

	ts.forEachChild(
		node,
		visitNode,
	);

	return meta;
}

function exilityTransformerFactory(context) {
	return function (file) {
		const imports = {};
		const result = visitNodeAndChildren(file, context, imports);

		return ts.updateSourceFileNode(
			result,
			Object.keys(imports)
				.map(name => imports[name])
				.concat(result.statements)
		);
	};
}

export default exilityTransformerFactory;
