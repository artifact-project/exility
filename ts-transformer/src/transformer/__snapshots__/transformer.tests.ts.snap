// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`arrow function 1`] = `
"import _exility_compile_helpers_src_stdlib_stdlib from \\"@exility/compile-helpers/src/stdlib/stdlib\\";
import _exility_dom_src_stddom_stddom from \\"@exility/dom/src/stddom/stddom\\";
class ArrowFunction extends Block<null> {
    static prototype.__template__ = (function templateFactory(__DEPS__) {
	\\"use strict\\";
	var __STDDOM = __DEPS__.stddom;
	var __STDDOM_CONTEXT = __STDDOM.CONTEXT;
	var __STDDOM_FRAGMENT = __STDDOM.FRAGMENT;
	var __STDDOM_NODE = __STDDOM.NODE;
	var __STDDOM_VALUE = __STDDOM.VALUE;
	var __STDDOM_TEXT = __STDDOM.TEXT;
	var __STDDOM_LIFECYCLE = __STDDOM.LIFECYCLE;
	var __STDDOM_UPD_VALUE = __STDDOM.UPD_VALUE;
	return function compiledTemplate(__SCOPE__) {
		__SCOPE__ = __SCOPE__ || {};
		var attrs = __SCOPE__.attrs;
		var __blocks__ = __SCOPE__.__blocks__;
		var __this__ = __SCOPE__.__this__;
		var __slots__ = __SCOPE__.__slots__;

		var __ctx = __STDDOM_CONTEXT();

		var __frag = __STDDOM_FRAGMENT();
		var __x1 = __STDDOM_NODE(__frag, \\"h1\\");
		__STDDOM_VALUE(__x1, __ctx, 2, attrs.text);
		__STDDOM_TEXT(__x1, \\"!\\");

		return {
			ctx: __ctx,
			frag: __frag,
			container: null,
			mountTo: function (container) {
				this.container = container;
				__frag.mountTo(container);
				__STDDOM_LIFECYCLE(__ctx, 'connectedCallback');
				return this;
			},
			update: function (__NEWSCOPE__) {
				__SCOPE__ = __NEWSCOPE__;
				attrs = __NEWSCOPE__.attrs;
				__blocks__ = __NEWSCOPE__.__blocks__;
				__this__ = __NEWSCOPE__.__this__;
				__slots__ = __NEWSCOPE__.__slots__;
				__STDDOM_UPD_VALUE(__ctx[2], attrs.text);
			}
		}
	}
})({ stdlib: stdlib_1.default, stddom: stddom_1.default });
}
"
`;

exports[`arrow function: for (item in items) 1`] = `
"import _exility_compile_helpers_src_stdlib_stdlib from \\"@exility/compile-helpers/src/stdlib/stdlib\\";
import _exility_dom_src_stddom_stddom from \\"@exility/dom/src/stddom/stddom\\";
class ArrowFor extends Block<null> {
    static prototype.__template__ = (function templateFactory(__DEPS__) {
	\\"use strict\\";
	var __STDDOM = __DEPS__.stddom;
	var __STDDOM_CONTEXT = __STDDOM.CONTEXT;
	var __STDDOM_FRAGMENT = __STDDOM.FRAGMENT;
	var __STDDOM_FOR = __STDDOM.FOR;
	var __STDDOM_NODE = __STDDOM.NODE;
	var __STDDOM_VALUE = __STDDOM.VALUE;
	var __STDDOM_UPD_VALUE = __STDDOM.UPD_VALUE;
	var __STDDOM_LIFECYCLE = __STDDOM.LIFECYCLE;
	var __STDDOM_UPD_FOR = __STDDOM.UPD_FOR;
	return function compiledTemplate(__SCOPE__) {
		__SCOPE__ = __SCOPE__ || {};
		var attrs = __SCOPE__.attrs;
		var __blocks__ = __SCOPE__.__blocks__;
		var __this__ = __SCOPE__.__this__;
		var __slots__ = __SCOPE__.__slots__;

		var __ctx = __STDDOM_CONTEXT();

		var __frag = __STDDOM_FRAGMENT();
		__STDDOM_FOR(__frag, __ctx, 2, attrs.items, undefined, __FOR_ITERATOR_1);

		function __FOR_ITERATOR_1(__parent, item, $index) {
			var __ctx = __STDDOM_CONTEXT(__parent);
			var __fragFor = __STDDOM_FRAGMENT();
			var __x3 = __STDDOM_NODE(__fragFor, \\"div\\");
			__STDDOM_VALUE(__x3, __ctx, 4, item);

			return {
				ctx: __ctx,
				frag: __fragFor,
				data: item,
				index: $index,
				update: function (__item, __$index) {
					item = __item
					$index = __$index
					__STDDOM_UPD_VALUE(__ctx[4], item);
				}
			};
		}

		return {
			ctx: __ctx,
			frag: __frag,
			container: null,
			mountTo: function (container) {
				this.container = container;
				__frag.mountTo(container);
				__STDDOM_LIFECYCLE(__ctx, 'connectedCallback');
				return this;
			},
			update: function (__NEWSCOPE__) {
				__SCOPE__ = __NEWSCOPE__;
				attrs = __NEWSCOPE__.attrs;
				__blocks__ = __NEWSCOPE__.__blocks__;
				__this__ = __NEWSCOPE__.__this__;
				__slots__ = __NEWSCOPE__.__slots__;
				__STDDOM_UPD_FOR(__ctx, 2, attrs.items, undefined, __FOR_ITERATOR_1);
			}
		}
	}
})({ stdlib: stdlib_1.default, stddom: stddom_1.default });
}
"
`;

exports[`arrow function: if 1`] = `
"import _exility_compile_helpers_src_stdlib_stdlib from \\"@exility/compile-helpers/src/stdlib/stdlib\\";
import _exility_dom_src_stddom_stddom from \\"@exility/dom/src/stddom/stddom\\";
class ArrowIf extends Block<null> {
    static prototype.__template__ = (function templateFactory(__DEPS__) {
	\\"use strict\\";
	var __STDDOM = __DEPS__.stddom;
	var __STDDOM_CONTEXT = __STDDOM.CONTEXT;
	var __STDDOM_FRAGMENT = __STDDOM.FRAGMENT;
	var __STDDOM_IF = __STDDOM.IF;
	var __STDDOM_NODE = __STDDOM.NODE;
	var __STDDOM_LIFECYCLE = __STDDOM.LIFECYCLE;
	var __STDDOM_UPD_IF = __STDDOM.UPD_IF;
	return function compiledTemplate(__SCOPE__) {
		__SCOPE__ = __SCOPE__ || {};
		var attrs = __SCOPE__.attrs;
		var __blocks__ = __SCOPE__.__blocks__;
		var __this__ = __SCOPE__.__this__;
		var __slots__ = __SCOPE__.__slots__;

		var __ctx = __STDDOM_CONTEXT();

		var __frag = __STDDOM_FRAGMENT();
		__STDDOM_IF(__frag, __ctx, 1, [__IF_2]);

		function __IF_2(frag) {
			return attrs.text ? frag || __IF_2_exec(__ctx) : null;
		}

		function __IF_2_exec(__parent) {
			var __ctx = __STDDOM_CONTEXT(__parent);
			var __fragIf = __STDDOM_FRAGMENT();
			var __x3 = __STDDOM_NODE(__fragIf, \\"h1\\");

			return {
				ctx: __ctx,
				frag: __fragIf,
				update: function () {
				}
			};
		}

		return {
			ctx: __ctx,
			frag: __frag,
			container: null,
			mountTo: function (container) {
				this.container = container;
				__frag.mountTo(container);
				__STDDOM_LIFECYCLE(__ctx, 'connectedCallback');
				return this;
			},
			update: function (__NEWSCOPE__) {
				__SCOPE__ = __NEWSCOPE__;
				attrs = __NEWSCOPE__.attrs;
				__blocks__ = __NEWSCOPE__.__blocks__;
				__this__ = __NEWSCOPE__.__this__;
				__slots__ = __NEWSCOPE__.__slots__;
				__STDDOM_UPD_IF(__ctx, 1);
			}
		}
	}
})({ stdlib: stdlib_1.default, stddom: stddom_1.default });
}
"
`;

exports[`blocks 1`] = `
"import _exility_compile_helpers_src_stdlib_stdlib from \\"@exility/compile-helpers/src/stdlib/stdlib\\";
import _exility_dom_src_stddom_stddom from \\"@exility/dom/src/stddom/stddom\\";
class App extends Block<{
        username: string;
    }> {
    static blocks = { Button };
    static prototype.__template__ = (function templateFactory(__DEPS__) {
	\\"use strict\\";
	var __STDDOM = __DEPS__.stddom;
	var __STDDOM_CMP_SET_COMPILER = __STDDOM.CMP_SET_COMPILER;
	var __STDDOM_CMP_INIT = __STDDOM.CMP_INIT;
	var __STDDOM_CONTEXT = __STDDOM.CONTEXT;
	var __STDDOM_FRAGMENT = __STDDOM.FRAGMENT;
	var __STDDOM_CMP_CREATE = __STDDOM.CMP_CREATE;
	var __STDDOM_LIFECYCLE = __STDDOM.LIFECYCLE;
	
	return function compiledTemplate(__SCOPE__) {
		__SCOPE__ = __SCOPE__ || {};
		var attrs = __SCOPE__.attrs;
		var __blocks__ = __SCOPE__.__blocks__;
		var __this__ = __SCOPE__.__this__;
		var __slots__ = __SCOPE__.__slots__;

		__STDDOM_CMP_INIT(__blocks__, [\\"Button\\"]);

		var __ctx = __STDDOM_CONTEXT();

		var __frag = __STDDOM_FRAGMENT();
		var __CMP_ATTRS = {};
		__CMP_ATTRS[\\"value\\"] = \\"Foo\\";
		var __x1 = __STDDOM_CMP_CREATE(__ctx, __blocks__,

			__frag,
			__this__,
			\\"Button\\",
			__CMP_ATTRS,
			null,
			null
		);

		return {
			ctx: __ctx,
			frag: __frag,
			container: null,
			mountTo: function (container) {
				this.container = container;
				__frag.mountTo(container);
				__STDDOM_LIFECYCLE(__ctx, 'connectedCallback');
				return this;
			},
			update: function (__NEWSCOPE__) {
				__SCOPE__ = __NEWSCOPE__;
				attrs = __NEWSCOPE__.attrs;
				__blocks__ = __NEWSCOPE__.__blocks__;
				__this__ = __NEWSCOPE__.__this__;
				__slots__ = __NEWSCOPE__.__slots__;
				var __CMP_ATTRS = {};
				__x1.update(__CMP_ATTRS);
			}
		}
	}
})({ stdlib: stdlib_1.default, stddom: stddom_1.default });
}
"
`;

exports[`classNames 1`] = `
"import _exility_compile_helpers_src_stdlib_stdlib from \\"@exility/compile-helpers/src/stdlib/stdlib\\";
import _exility_dom_src_stddom_stddom from \\"@exility/dom/src/stddom/stddom\\";
class App extends Block<{
        username: string;
    }> {
    static classNames = css;
    static prototype.__template__ = (function templateFactory(__DEPS__) {
	\\"use strict\\";
	var __STDDOM = __DEPS__.stddom;
	var __STDDOM_CONTEXT = __STDDOM.CONTEXT;
	var __STDDOM_FRAGMENT = __STDDOM.FRAGMENT;
	var __STDDOM_LIVE_NODE = __STDDOM.LIVE_NODE;
	var __STDDOM_D_PROP = __STDDOM.D_PROP;
	var __STDDOM_TEXT = __STDDOM.TEXT;
	var __STDDOM_LIFECYCLE = __STDDOM.LIFECYCLE;
	var __STDLIB = __DEPS__.stdlib;
	var __STDLIB_CSS_MODULE = __STDLIB.CSS_MODULE;
	return function compiledTemplate(__SCOPE__) {
		__SCOPE__ = __SCOPE__ || {};
		var attrs = __SCOPE__.attrs;
		var __blocks__ = __SCOPE__.__blocks__;
		var __this__ = __SCOPE__.__this__;
		var __slots__ = __SCOPE__.__slots__;
		var __classNames__ = __SCOPE__.__classNames__;

		var __CLN = __STDLIB_CSS_MODULE(__classNames__);

		var __ctx = __STDDOM_CONTEXT();

		var __frag = __STDDOM_FRAGMENT();
		var __x1 = __STDDOM_LIVE_NODE(__frag, __ctx, 1, \\"div\\");
		__STDDOM_D_PROP(__ctx[1], \\"className\\", __CLN(\\"foo\\"))
		__STDDOM_TEXT(__x1, \\"bar\\");

		return {
			ctx: __ctx,
			frag: __frag,
			container: null,
			mountTo: function (container) {
				this.container = container;
				__frag.mountTo(container);
				__STDDOM_LIFECYCLE(__ctx, 'connectedCallback');
				return this;
			},
			update: function (__NEWSCOPE__) {
				__SCOPE__ = __NEWSCOPE__;
				attrs = __NEWSCOPE__.attrs;
				__blocks__ = __NEWSCOPE__.__blocks__;
				__this__ = __NEWSCOPE__.__this__;
				__slots__ = __NEWSCOPE__.__slots__;
				__classNames__ = __NEWSCOPE__.__classNames__;
				__CLN.upd(__classNames__);
				__STDDOM_D_PROP(__ctx[1], \\"className\\", __CLN(\\"foo\\"))
			}
		}
	}
})({ stdlib: stdlib_1.default, stddom: stddom_1.default });
}
"
`;

exports[`string 1`] = `
"import _exility_compile_helpers_src_stdlib_stdlib from \\"@exility/compile-helpers/src/stdlib/stdlib\\";
import _exility_dom_src_stddom_stddom from \\"@exility/dom/src/stddom/stddom\\";
class Str extends Block {
    static prototype.__template__ = (function templateFactory(__DEPS__) {
	\\"use strict\\";
	var __STDDOM = __DEPS__.stddom;
	var __STDDOM_CONTEXT = __STDDOM.CONTEXT;
	var __STDDOM_FRAGMENT = __STDDOM.FRAGMENT;
	var __STDDOM_NODE = __STDDOM.NODE;
	var __STDDOM_VALUE = __STDDOM.VALUE;
	var __STDDOM_TEXT = __STDDOM.TEXT;
	var __STDDOM_LIFECYCLE = __STDDOM.LIFECYCLE;
	var __STDDOM_UPD_VALUE = __STDDOM.UPD_VALUE;
	return function compiledTemplate(__SCOPE__) {
		__SCOPE__ = __SCOPE__ || {};
		var attrs = __SCOPE__.attrs;
		var __blocks__ = __SCOPE__.__blocks__;
		var __this__ = __SCOPE__.__this__;
		var __slots__ = __SCOPE__.__slots__;

		var __ctx = __STDDOM_CONTEXT();

		var __frag = __STDDOM_FRAGMENT();
		var __x1 = __STDDOM_NODE(__frag, \\"h1\\");
		__STDDOM_VALUE(__x1, __ctx, 2, attrs.text);
		__STDDOM_TEXT(__x1, \\"!\\");

		return {
			ctx: __ctx,
			frag: __frag,
			container: null,
			mountTo: function (container) {
				this.container = container;
				__frag.mountTo(container);
				__STDDOM_LIFECYCLE(__ctx, 'connectedCallback');
				return this;
			},
			update: function (__NEWSCOPE__) {
				__SCOPE__ = __NEWSCOPE__;
				attrs = __NEWSCOPE__.attrs;
				__blocks__ = __NEWSCOPE__.__blocks__;
				__this__ = __NEWSCOPE__.__this__;
				__slots__ = __NEWSCOPE__.__slots__;
				__STDDOM_UPD_VALUE(__ctx[2], attrs.text);
			}
		}
	}
})({ stdlib: stdlib_1.default, stddom: stddom_1.default });
}
"
`;
